XenAdmin/Wizards/RollingUpgradeWizard/PlanActions/RpuUploadAndApplySuppPackPlanAction.cs

```diff
@@ -0,0 +1,267 @@
+﻿/* Copyright (c) Citrix Systems, Inc. 
+ * All rights reserved. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using XenAdmin.Actions;
+using XenAdmin.Core;
+using XenAPI;
+using XenAdmin.Diagnostics.Checks;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+using XenAdmin.Network;
+using XenAdmin.Wizards.PatchingWizard;
+using XenAdmin.Wizards.PatchingWizard.PlanActions;
+
+
+namespace XenAdmin.Wizards.RollingUpgradeWizard.PlanActions
+{
+    class RpuUploadAndApplySuppPackPlanAction : PlanActionWithSession
+    {
+        private Host host;
+        private List<Host> hosts;
+        private string suppPackPath;
+        private Dictionary<Host, Pool_update> uploadedSuppPacks;
+        public List<PlanAction> delayedPlanActions;

```

Comment:
Storing the delayed plan actions on the RpuUploadAndApplySuppPackPlanAction is not optimum design and is a bit inconsistent with the other plan actions. Beyond the class design though, we have a bigger problem: these delayed actions are populated when the RpuUploadAndApplySuppPackPlanAction actually runs; if the delayed action is RestartHost, we get the VMs wrong, because by this time the host has been evacuated already. I think we should be able to calculate the delayed actions and add them to the greater list of delayed actions in the RollingUpgradeUpgradePage at the time RpuUploadAndApplySuppPackPlanAction  is constructed. Thus you can use the parent class's method getting the after-apply guidance instead of duplicating the code here.